### ISSUE TITLE
💅 useRegexpExec is 3.65x slower in 2.5.12 on a Zod enum call

### ISSUE THREAD

**Issue by jakeleventhal** [biomejs/biome#11610]
### Environment information

<details>

```text
CLI:
  Version: 2.5.12

Platform:
  CPU Architecture: aarch64
  OS: macos 27.0
  CPU: Apple M1 Max

Environment:
  BIOME_DISTRIBUTION: npm
  JS_RUNTIME_VERSION: v24.12.0
  JS_RUNTIME_NAME: node

Biome Configuration:
  Status: Loaded successfully
  Path: biome.json
  VCS enabled: false
  Enabled rule: nursery/useRegexpExec
```

</details>

### Rule name

`nursery/useRegexpExec`

### Playground link

https://github.com/jakeleventhal/biome-use-regexp-exec-perf-repro

The repository pins Biome 2.5.10 and 2.5.12 side by side. Run:

```sh
bun install --frozen-lockfile
bun run bench
bun run profile:2.5.10
bun run profile:2.5.12
```

The analyzed file contains one call expression:

```ts
import { schema } from "./schema";

schema.safeParse("one");
```

`schema` is a two-value Zod 4.3.6 enum.

Five-run averages on the machine above:

| Version | End-to-end average | `useRegexpExec` rule time |
| --- | ---: | ---: |
| 2.5.10 | 1.830s | 1.484s |
| 2.5.12 | 5.877s | 5.417s |

That is a 3.21x end-to-end slowdown and a 3.65x increase inside the single rule invocation.

### Expected result

Biome 2.5.12 should process this single call in roughly the same time as 2.5.10. `useRegexpExec` should not spend more than five seconds resolving the type of one `ZodEnum.safeParse()` call.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct
